### PR TITLE
[GL-1597] HCA Optimus Adapters

### PR DESCRIPTION
**Pull request by cheyenne-gold**

### CODE REVIEW

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    
+    # Create the project level objects based on the intermediate looms and the final merged loom
+    call CreateOptimusObjects.CreateOptimusAdapterObjects as CreateProjectOptimusAdapters {
+      input:
+        loom = MergeLooms.project_loom,
+        process_input_ids = [GetProjectLevelInputIds.process_input_uuids],
+        input_id = project_stratum_string,
+        library = library,
```
Comment (wjdingman1) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:162]: See comment above

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+import "../../projects/tasks/AdapterTasks.wdl" as Tasks
+import "../../projects/tasks/CreateReferenceMetadata.wdl" as CreateReferenceMetadata
+
+
+workflow CreateAdapterMetadata {
+  meta {
+    description: "Creates json objects for indexing HCA analysis data"
+    allowNestedInputs: true
+  }
+
```
Comment (wjdingman1) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:14]: Maybe make a comments about fastqs being an array of space separated strings...our future selves will be thankful

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  input {
+    Array[File] output_bams
+    Array[File] output_looms
+    Array[File]? output_bais
```
Comment (wjdingman1) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:18]: can probably get rid out output_bais

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    Boolean is_project_level
+    String? pipeline_version # parsed from metadata for intermediate, passed in for project level
+    String? reference_file_fasta # parsed from metadata for intermediate, passed in for project level
+  }
+
+  call Tasks.GetCromwellMetadata {
+    input:
+      output_path = loom,
+      cromwell_url = cromwell_url,
+      include_subworkflows = false # TODO: do we need subworkflows???
```
Comment (wjdingman1) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:35]: remove comment

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    # store variable resulting from project run
+    Array[File] project_links = CreateProjectOptimusAdapters.links_outputs
+    Array[File] project_analysis_process_objects = CreateProjectOptimusAdapters.analysis_process_outputs
+    Array[File] project_analysis_protocol_objects = CreateProjectOptimusAdapters.analysis_protocol_outputs
+    Array[File] project_analysis_file_objects = CreateProjectOptimusAdapters.analysis_file_outputs
+    Array[File] project_loom_descriptor_objects = CreateProjectOptimusAdapters.loom_file_descriptor_outputs
+
+  #}
+
+  ########################## Get SS2 Metadata Files ###########################
```
Comment (wjdingman1) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:184]: remove comment block

File: projects/tasks/AdapterTasks.wdl [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+    String project_id
+    Array[String] process_input_ids
+    File output_file_path
+    String version_timestamp
+    Array[File] analysis_process_path
+    Array[File] analysis_protocol_path
+    String file_name_string
+    Boolean project_level
+
+    String docker = "us.gcr.io/broad-gotc-prod/pipeline-tools:latest" #remember to change back to quay?
```
Comment (wjdingman1) [inline projects/tasks/AdapterTasks.wdl:299]: remove comment

**Review by wjdingman1 — APPROVED** [review]

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    Array[String] all_libraries
+    Array[String] all_organs
+    Array[String] all_species
+    Array[String] all_project_ids
+    Array[String] all_project_names
+
+    String output_basename
+    String cromwell_url = "https://api.firecloud.org/"
+    String staging_area = "gs://broad-dsp-monster-hca-prod-lantern/"
+    String version_timestamp = "2021-05-24T12:00:00.000000Z" # TODO should we hard code this?
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:33]:
```suggestion
    String version_timestamp
```

Comment (cheyenne-gold) [reply projects/hca_mvp/CreateAdapterMetadata.wdl:33]: Ask if anyone has objections to this at DCP demos on 8/31. Timestamp should always be left at midnight (12:00:00.000000Z). Add this to spreadsheet/ask Marina to copy it in on SOP

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+  Array[File] intermediate_analysis_protocol_objects = flatten(CreateIntermediateOptimusAdapters.analysis_protocol_outputs)
+  Array[File] intermediate_analysis_file_objects = flatten(CreateIntermediateOptimusAdapters.analysis_file_outputs)
+  Array[File] intermediate_loom_descriptor_objects = flatten(CreateIntermediateOptimusAdapters.loom_file_descriptor_outputs)
+  Array[File] intermediate_bam_descriptor_objects = flatten(select_all(CreateIntermediateOptimusAdapters.bam_file_descriptor_outputs))
+
+  call CreateReferenceMetadata.CreateReferenceMetadata as CreateReferenceMetadata {
+    input:
+      reference_fastas = CreateIntermediateOptimusAdapters.reference_fasta,
+      species = species,
+      pipeline_type = 'Optimus',
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:123]:
```suggestion
      pipeline_type = "Optimus",
```

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+  # store variable resulting from intermediate run
+  Array[File] intermediate_links = flatten(CreateIntermediateOptimusAdapters.links_outputs)
+  Array[File] intermediate_analysis_process_objects = flatten(CreateIntermediateOptimusAdapters.analysis_process_outputs)
+  Array[File] intermediate_analysis_protocol_objects = flatten(CreateIntermediateOptimusAdapters.analysis_protocol_outputs)
+  Array[File] intermediate_analysis_file_objects = flatten(CreateIntermediateOptimusAdapters.analysis_file_outputs)
+  Array[File] intermediate_loom_descriptor_objects = flatten(CreateIntermediateOptimusAdapters.loom_file_descriptor_outputs)
+  Array[File] intermediate_bam_descriptor_objects = flatten(select_all(CreateIntermediateOptimusAdapters.bam_file_descriptor_outputs))
+
+  call CreateReferenceMetadata.CreateReferenceMetadata as CreateReferenceMetadata {
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:119]:
```suggestion
  call CreateReferenceMetadata.CreateReferenceMetadata {
```

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+      species = species,
+      pipeline_type = 'Optimus',
+      version_timestamp = version_timestamp,
+      input_type = "reference"
+  }
+
+  Array[File] reference_fasta_array = [CreateReferenceMetadata.reference_fasta]
+
+  # Merge all intermediate run looms to a single project level loom
+  call MergeLooms.MergeOptimusLooms as MergeLooms {
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:131]:
```suggestion
  call MergeLooms.MergeOptimusLooms {
```

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+      output_looms = output_looms,
+      library = library,
+      species = species,
+      organ = organ,
+      project_id = project_id,
+      project_name = project_name,
+      output_basename = output_basename
+  }
+
+  Array[File] project_loom_array = [MergeLooms.project_loom]
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:142]:
```suggestion
  Array[File] project_loom_array = [MergeOptimusLooms.project_loom]
```

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  call Tasks.GetProjectLevelInputIds {
+    input:
+      intermediate_analysis_files = flatten(CreateIntermediateOptimusAdapters.analysis_file_outputs)
+  }
+
+  
+  # Create the project level objects based on the intermediate looms and the final merged loom
+  call CreateOptimusObjects.CreateOptimusAdapterObjects as CreateProjectOptimusAdapters {
+    input:
+      loom = MergeLooms.project_loom,
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:154]:
```suggestion
      loom = MergeOptimusLooms.project_loom,
```

File: projects/hca_mvp/CreateAdapterMetadata.wdl [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+      process_input_ids = [GetProjectLevelInputIds.process_input_uuids],
+      input_id = project_stratum_string,
+      project_id = project_id,
+      version_timestamp = version_timestamp,
+      cromwell_url = cromwell_url,
+      is_project_level = true,
+      reference_file_fasta = CreateIntermediateOptimusAdapters.reference_fasta[0],
+      pipeline_version = CreateIntermediateOptimusAdapters.pipeline_version_string[0]
```
Comment (cheyenne-gold) [inline projects/hca_mvp/CreateAdapterMetadata.wdl:162]:
```suggestion
      pipeline_version = MergeOptimusLooms.pipeline_version_string
```

File: projects/tasks/AdapterTasks.wdl [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+}
+
+task GetPipelineVersion {
+  input {
+    String pipeline_version
+
+    String docker = "ubuntu:18.04"
+  }
+  command {
+    echo ~{pipeline_version} > pipeline_version.txt
```
Comment (cheyenne-gold) [inline projects/tasks/AdapterTasks.wdl:568]:
```suggestion
        echo ~{prefix}~{pipeline_version} > pipeline_version.txt
```

File: projects/tasks/AdapterTasks.wdl [hunk 15]
```diff
@@ -0,0 +1,579 @@
...
+    cpu: cpu
+    memory: "~{memory} GiB"
+    disks: "local-disk ~{disk} HDD"
+  }
+}
+
+task GetPipelineVersion {
+  input {
+    String pipeline_version
+
```
Comment (cheyenne-gold) [inline projects/tasks/AdapterTasks.wdl:564]:
```suggestion
    String prefix = "MergeOptimusLooms_v"
```

File: projects/tasks/AdapterTasks.wdl [hunk 16]
```diff
@@ -0,0 +1,579 @@
...
+task GetAnalysisProcessMetadata {
+  input {
+    String input_uuid
+    String pipeline_type
+    String version_timestamp
+    String references
+    File input_file
+    Boolean project_level
+    String? loom_timestamp
```
Comment (cheyenne-gold) [inline projects/tasks/AdapterTasks.wdl:223]:
```suggestion
```

File: projects/tasks/AdapterTasks.wdl [hunk 17]
```diff
@@ -0,0 +1,579 @@
...
+
+  command {
+    create-analysis-process \
+      --input_uuid "~{input_uuid}" \
+      --pipeline_type "~{pipeline_type}" \
+      --workspace_version "~{version_timestamp}" \
+      --references "~{references}" \
+      --input_file "~{input_file}" \
+      --project_level ~{project_level} \
+      ~{"--loom_timestamp " + loom_timestamp}
```
Comment (cheyenne-gold) [inline projects/tasks/AdapterTasks.wdl:239]:
```suggestion
```

File: projects/tasks/AdapterTasks.wdl [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+
+task ParseCromwellMetadata {
+  input {
+    File cromwell_metadata
+    String pipeline_type
+
+    String docker = "us.gcr.io/broad-gotc-prod/pipeline-tools:latest"
+    Int cpu = 1
+    Int machine_mem_mb = 2000
+    Int disk = 10
```
Comment (cheyenne-gold) [inline projects/tasks/AdapterTasks.wdl:446]:
```suggestion
    Int disk = ceil((size(cromwell_metadata, "G") * 2)) + 5
```

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+      project_level = is_project_level,
+      project_loom = loom
+  }
+
+  call Tasks.GetAnalysisProcessMetadata {
+    input:
+      input_uuid = input_id,
+      pipeline_type = pipeline_type,
+      version_timestamp = version_timestamp,
+      references = select_first([ParseCromwellMetadata.ref_fasta,reference_file_fasta]),
```
Comment (cheyenne-gold) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:59]:
```suggestion
      references = reference,
```

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+      input_file = GetCromwellMetadata.metadata
+  }
+
+  call Tasks.GetAnalysisProtocolMetadata {
+    input:
+      input_uuid = input_id,
+      pipeline_type = pipeline_type,
+      version_timestamp = version_timestamp,
+      project_level = is_project_level,
+      pipeline_version = select_first([ParseCromwellMetadata.pipeline_version,pipeline_version])
```
Comment (cheyenne-gold) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:70]:
```suggestion
      pipeline_version = pipe_version
```

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  call Tasks.GetCloudFileCreationDate  as GetLoomFileCreationDate {
```
Comment (cheyenne-gold) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:73]:
```suggestion
  call Tasks.GetCloudFileCreationDate as GetLoomFileCreationDate {
```

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        file_path = select_first([bam])
+    }
+
+    call Tasks.GetFileDescriptor as GetBamFileDescriptor {
+      input:
+        pipeline_type = pipeline_type,
+        file_path = select_first([bam]),
+        input_uuid = input_id,
+        creation_time = GetBamFileCreationDate.creation_date,
+        version_timestamp = version_timestamp,
```
Comment (cheyenne-gold) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:99]:
```suggestion
        version_timestamp = version_timestamp
```

File: projects/tasks/CreateOptimusAdapterObjects.wdl [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+      file_path = loom
+  }
+
+  call Tasks.GetFileDescriptor as GetLoomFileDescriptor {
+    input:
+      pipeline_type = pipeline_type,
+      file_path = loom,
+      input_uuid = input_id,
+      creation_time = GetLoomFileCreationDate.creation_date,
+      version_timestamp = version_timestamp,
```
Comment (cheyenne-gold) [inline projects/tasks/CreateOptimusAdapterObjects.wdl:84]:
```suggestion
      version_timestamp = version_timestamp
```

**Review by jessicaway — APPROVED** [review]
It's beautiful!

**Pull request merged**